go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+* Currently sending is sent separate and async, so the current version doesn't wrap the handling. Instead this will be
+  part of the async call and documentation will make clear it needs to be safe for concurrent invocation.
+* The Java version provides the event ID so we will too
+
+### WorfkflowInboundCallsInterceptor.HandleQuery
+
+Current:
+
+```go
+HandleQuery(ctx Context, queryType string, args *commonpb.Payloads,
+  handler func(*commonpb.Payloads) (*commonpb.Payloads, error)) (*commonpb.Payloads, error)
+```
+
+Proposal:
+
+```go
+type QueryInput struct {
+  QueryType string
+  Args []interface{}
+}
+
+HandleQuery(ctx Context, input *QueryInput) (interface{}, error)
+```
+
+Why:
+
+* All other calls convert the payload outside of the interceptor call, this can too
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
+
+### WorkflowOutboundCallsInterceptor
+
+We can leave everything the way it is because, while it doesn't make the interceptor parameters future-proof with struct
+wrappers, it mimics the actual calls that aren't going to have backwards-compatible alterations.
+
+### WorkflowOutboundCallsInterceptor.GetWorkflowInfo
+
+Current:
+
+```go
+GetWorkflowInfo(ctx Context) *WorkflowInfo
+```
+
+Proposal:
+
+```go
+GetInfo(ctx Context) *WorkflowInfo
+```
+
+Why: All other calls match `workflow` package, why not this?
+
+### ActivityInterceptor
+
+Proposal:
+
+```go
+type ActivityInterceptor interface {
+  InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+}
+```
+
+Why: Mimics `WorkflowInterceptor` but for activities.
+
+### ActivityInboundInterceptor
+
+Proposal:
+
+```go
+type ActivityInput struct {
+  Args []interface{}
+}
+
+type ActivityInboundInterceptor interface {
+  Init(outbound ActivityOutboundInterceptor) error
+
+  ExecuteActivity(ctx context.Context, input *ActivityInput) (interface{}, error)
+}
+```
+
+Also a `ActivityInboundInterceptorBase` implementation with expected delegation.
+
+Why: Mimics `WorkflowInboundInterceptor` but for activities.
+
+### ActivityOutboundInterceptor
+
+Proposal:
+
+```go
+type ActivityOutboundInterceptor interface {
+  GetInfo(ctx context.Context) ActivityInfo
+  GetLogger(ctx Context) log.Logger
+  GetMetricsScope(ctx Context) tally.Scope
+  RecordHeartbeat(ctx context.Context, details ...interface{})
+  HasHeartbeatDetails(ctx context.Context) bool
+  GetHeartbeatDetails(ctx context.Context, d ...interface{}) error
+  GetWorkerStopChannel(ctx context.Context) <-chan struct{}
+}
+```
+
+Also a `ActivityOutboundInterceptorBase` implementation with expected delegation.
+
+Why:
+
+* Mimics `WorkflowOutboundInterceptor` but for activities.
+
+For discussion: `GetInfo` returns expensive copies because `activity.GetInfo` does. Is this ok? This is inconsistent
+with workflow info which is a pointer.
+
+## Interceptor Possible Updates
+
+### Forced Base Embedding
```

Comment:
Do we want this?

Reply:
I don't totally follow what this does. Can you gimme a little explainer on how this helps with compat?